### PR TITLE
Fix keyless verification of openvex attesatations

### LINKED ISSUE #143: `vexctl attest --sign` doesn't create a verifiable attestation
# Description

When calling `vexctl attest --sign --attach report.vex.json`, the attached attestation seemingly cannot be verified via `cosign verify-attestion`.

## Repro steps:
1. Create a sample report with some data:
```
{
  "@context": "https://openvex.dev/ns/v0.2.0",
  "@id": "https://openvex.dev/docs/public/vex-f7c629e617119c33512cc55395988295b80d3a450266c8399927a6fec323520d",
  "author": "dhaus",
  "timestamp": "2023-12-05T03:51:30.679763+01:00",
  "version": 1,
  "statements": [
    {
      "vulnerability": {
        "name": "CVE-2022-1304"
      },
      "timestamp": "2023-12-05T03:51:30.679765+01:00",
      "products": [
        {
          "@id": "pkg:oci/openvex@sha256:a97a153152fcd6410bdf4fb64f5622ecf97a753f07dcc89dab14509d059736cf?repository_url=docker.io/daha97&tag=1.23.4"
        }
      ],
      "status": "not_affected",
      "justification": "vulnerable_code_not_in_execute_path",
      "impact_statement": "nothing"
    }
  ]
}
```

2. Attach and sign the report going through the signing flow
```
vexctl attest --attach --sign vex.json

Generating ephemeral keys...
Retrieving signed certificate...
...
Successfully verified SCT...
```

3. Attempt to verify the attestation
```
cosign verify-attestation docker.io/daha97/openvex:1.23.4 --certificate-identity="..." --certificate-oidc-issuer=https://github.com/login/oauth

Error: no matching attestations: nil certificate provided
main.go:74: error during command execution: no matching attestations: nil certificate provided
```

Looking at the manifest attestation in the registry, no additional annotations are set:
```
{
    "mediaType": "application/vnd.dsse.envelope.v1+json",
    "digest": "sha256:49709de8de52ec1ecb610900e2066e557edba647d70944461caaba47cf14b4dd",
    "size": 1724,
    "annotations": {
        "dev.cosignproject.cosign/signature": ""
    }
}
```

Workaround for this:

Creating the vex report but using `cosign` to attach the report instead of `vexctl attest --attach` works:
```
cosign attest --type custom --predicate vex.json docker.io/daha97/openvex:1.23.4

Generating ephemeral keys...
Retrieving signed certificate...
Successfully verified SCT...

cosign verify-attestation docker.io/daha97/openvex:1.23.4 --certificate-identity="..." --certificate-oidc-issuer=https://github.com/login/oauth
```

Also, the annotations are now correctly set:
```
{
    "mediaType": "application/vnd.dsse.envelope.v1+json",
    "digest": "sha256:50375613538f8b551ed758fb014d0832cdb1343f18b94bbc627e7d01cb7fc40c",
    "size": 1684,
    "annotations": {
        "predicateType": "https://cosign.sigstore.dev/attestation/v1",
        "dev.cosignproject.cosign/signature": "",
        "dev.sigstore.cosign/bundle": "...",
        "dev.sigstore.cosign/certificate": "...",
        "dev.sigstore.cosign/chain": "..."
    }
}
```

Sample attestation manifest containing both predicates: https://oci.dag.dev/?image=daha97%2Fopenvex%3Asha256-a97a153152fcd6410bdf4fb64f5622ecf97a753f07dcc89dab14509d059736cf.att

Maybe I'm also doing something wrong, but seemingly cannot get it to work with `vexctl attest --attach --sign`.

### PR DESCRIPTION
**Pull request by puerco**
### TLDR

This is a large PR that fixes keyless verification of the OpenVEX attestations as reported in https://github.com/openvex/vexctl/issues/143 

This is a solid but temporary fix, the functionality in this PR will be moved to `openvex/discoery` and an upcoming trust module but it should do the work while we finish those new modules.

**The summary of changes**
1. We now upload the signatures to rekor
2. Results of signing operation and appending to the tlog are now recorded in the attestation struct
3. We now add the needed  OCI annotations when attaching images to enable keyless verification

### Commit 4e44f2164899c1a7e38880c79d185ec79eb236d3: Refactor attestation.Sign() + Tlog append

The main goal is to add two missing features:

1. Register the signature data to Rekor<br>
After signing, we now register the signature in the sigstore transparency
log. This is essential to allow for keyless verification.

2. New SignatureData Field<br>
The attestation now has a new SignatureData field that captures the results
of the signing operation. This is required to make data like the cert and the
proof of inclusion available externally (eg to record them in oci annotations).

The `attestation.Sign()` method has been heavily refactored but should be simppler
as the work it does is now broken into three internal functions:

*  `initSigning`: creates context and options
*  `signAttestation`: Performs the actual signing
*  `appendSignatureDataToTLog`: Uploads data to rekor

### Commit 5aa0d4485f6e8c38e88456fa35e3b08aca626e12: OCI Annotations

This commit modifies the attachAttestation function of the vexctl
implementation to add the OCI annotations required to keylessly
verify OpenVEX attestations.

/cc @cpanato 
Fixes #143 

Signed-off-by: Adolfo García Veytia (Puerco) <puerco@chainguard.dev>

### CODE REVIEW

**Comment by puerco** [comment]
cosign verify attestation now works

image

**Comment by cpanato** [comment]
thanks
